src/filter.rs

```diff
@@ -105,152 +105,90 @@ pub fn separable_filter<I, K>(image: &I, h_kernel: &[K], v_kernel: &[K])
 pub fn separable_filter_equal<I, K>(image: &I, kernel: &[K])
         -> VecBuffer<I::Pixel>
     where I: GenericImage,
-          I::Pixel: HasBlack + 'static,
+          I::Pixel: 'static,
           <I::Pixel as Pixel>::Subpixel: ValueInto<K> + Clamp<K>,
           K: Num + Copy {
     separable_filter(image, kernel, kernel)
 }
 
-/// Returns 2d correlation of an image with a 3x3 kernel. Intermediate calculations are
-/// performed at type K, and the results clamped to subpixel type S. Pads by continuity.
-// TODO: factor out the accumulation code from this, horizontal_filter and vertical_filter
-pub fn filter3x3<I, P, K, S>(image: &I, kernel: &[K]) -> VecBuffer<ChannelMap<P, S>>
-    where I: GenericImage<Pixel=P>,
-          P::Subpixel: ValueInto<K>,
-          S: Clamp<K> + Primitive + 'static,
-          P: WithChannel<S> + 'static,
+/// Returns 2d correlation of an image with a kernel. Intermediate calculations are
+/// performed at type K, and the results converted to pixel Q via f. Pads by continuity.
+pub fn filter_kernel<I, K, F, Q>(image: &I, kernel: &[K], 
+                                 k_width: u32, k_height: u32, mut f: F) 
+        -> VecBuffer<Q>
+    where I: GenericImage,
+          <I::Pixel as Pixel>::Subpixel: ValueInto<K>,
+          Q: Pixel + 'static,
+          F: FnMut(&mut Q::Subpixel, K) -> (),
           K: Num + Copy {
 
-    let (width, height) = image.dimensions();
-    let mut out = ImageBuffer::<ChannelMap<P, S>, Vec<S>>::new(width, height);
-    let num_channels = I::Pixel::channel_count() as usize;
+    assert!(kernel.len() as u32 >= k_width * k_height,
+            "incompatible kernel lengths".to_owned());
 
```

Comment:
It doesn't feel right to have to pass in width and height separately from the kernel. What's the use case for allowing kernel length > width * height? 

We could introduce a trivial Kernel struct just containing an array (slice?), width and height.